app/src/main/java/org/kiwix/kiwixmobile/wifi_hotspot/HotspotService.java

```diff
@@ -0,0 +1,200 @@
+        break;
+
+      case ACTION_TURN_ON_AFTER_O:
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          //serverStateListener.hotspotTurnedOn(hotspotManager.turnOnHotspot());
+          hotspotManager.turnOnHotspot(serverStateListener);
+          updateNotification(getString(R.string.hotspot_running), true);
+        }
+        break;
+
+      case ACTION_TURN_OFF_AFTER_O:
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+          stopHotspot();
+        }
+        break;
+
+      case ACTION_START_SERVER:
+        webServerHelper.startServerHelper(serverStateListener);
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+          updateNotification(getString(R.string.hotspot_running), true);
+        }
+        break;
+
+      case ACTION_STOP_SERVER:
+        dismissNotification();
+        break;
+      default:
+        break;
+    }
+    return START_NOT_STICKY;
+  }
+
+  @Nullable @Override public IBinder onBind(Intent intent) {
+    return serviceBinder;
+  }
+
+  private Notification buildForegroundNotification(String status, boolean showStopButton) {
+    Log.v(TAG, "Building notification " + status);
+    builder = new NotificationCompat.Builder(this);
+    builder.setContentTitle("Kiwix Hotspot").setContentText(status);
+    Intent targetIntent = new Intent(this, ZimHostActivity.class);
+    targetIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+    PendingIntent contentIntent =
+        PendingIntent.getActivity(this, 0, targetIntent, PendingIntent.FLAG_UPDATE_CURRENT);
+    builder.setContentIntent(contentIntent)
+        .setSmallIcon(R.mipmap.kiwix_icon)
+        .setWhen(System.currentTimeMillis());
+
+    hotspotNotificationChannel();
+
+    if (showStopButton) {
+      Intent stopIntent = new Intent(ACTION_STOP);
+      PendingIntent stopHotspot =
+          PendingIntent.getBroadcast(this, 0, stopIntent, PendingIntent.FLAG_UPDATE_CURRENT);
+      builder.addAction(R.drawable.ic_close_white_24dp, getString(R.string.stop_hotspot_button),
+          stopHotspot);
+    }
+    return (builder.build());
+  }
+
+  private void updateNotification(String status, boolean stopAction) {
+    notificationManager.notify(HOTSPOT_NOTIFICATION_ID,
+        buildForegroundNotification(status, stopAction));
+  }
+
+  //Dismiss notification and turn off hotspot for devices>=O
+  @RequiresApi(Build.VERSION_CODES.O)
+  void stopHotspot() {
+    hotspotManager.turnOffHotspot();
+    webServerHelper.stopAndroidWebServer(serverStateListener);
+    stopForeground(true);
+    stopSelf();
+  }
+
+  //Dismiss notification and turn off hotspot for devices < O
+  void dismissNotification() {
+    webServerHelper.stopAndroidWebServer(serverStateListener);
+    stopForeground(true);
+    stopSelf();
+  }
+
+  @Override
+  public void onDestroy() {
+    if (stopReceiver != null) {
+      unregisterReceiver(stopReceiver);
+    }
+    super.onDestroy();
+  }
+
+  private void hotspotNotificationChannel() {
```

Comment:
I'd like to see all the clutter in this class around Notifications extracted into a new class. KiwixNotificationManager perhaps.

Also NotificationChannels are a big topic but we need to make sure we delete old ones and create new ones across versions, they can be a hell of a nuisance to maintain overtime and the choice of when to create them is also vital as generally you want to do it as early as possible so that whenever the user visits our settings screen the can see all the types of notifications we have. Unless we do want our channels to be dynamic. It is a bit of a minefield is my point

Reply:
Sure. I will look into this.